### DISCUSSION TITLE
Debugging the firmware

### DISCUSSION THREAD

**Discussion by Novakasa** [pybricks/support#629] [Q&A] [answered]
Since I am not an expert developing with C let alone developing in micropython, I am wondering what are some approaches to debugging when developing the firmware. I am guessing we can call the micropython print function in C somehow. Are there some other tricks I might benefit from?

**Comment by dlech** [comment, marked as answer]
Debug printing is about as good as it gets for code running on the hub and sometimes that isn't even possible (https://github.com/pybricks/pybricks-micropython/pull/80#issuecomment-1008333124).

For code that doesn't require actual hardware, the best bet is to write unit tests that can run on the host computer instead. Then you can actually build a debug build and attach a debugger if needed. Also the unit tests will help prevent any future regressions of the code. For code that doesn't actually require MicroPython, the code and tests should go in `lib/pbio`. For MicroPython code, we don't have a good solution yet but we are considering implementing a "virtual" hub that would allow running Pybricks MicroPython on the host computer.

**Reply by Novakasa** [reply]
Thanks. The approach with unit tests is definitely one I will use. What purpose has the "debug" build config? I am noticing that CI fails with my PR due to firmware size of the "debug firmware".

**Reply by dlech** [reply]
I have a STM32 development board that I can actually attach a debugger to. Hopefully the setting `FIXMATH_NO_CACHE` will fix that. Since it uses a debug build it probably doesn't remove unused symbols.

**Comment by massimiliano-mantione** [comment]
Is there a way to attach a debug probe to a technic hub board?
What is the test setup [here](https://github.com/pybricks/support/issues/1095#issuecomment-1783899701) for?

More generally: what did you use to bootstrap pybricks on the board?
Just UART-based debugging?

**Reply by dlech** [reply]
> Is there a way to attach a debug probe to a technic hub board?

The STM32 chip is locked at the factory, so we can't attach a hardware debugger. I haven't tried with the Bluetooth chip though.

> What is the test setup [here](https://github.com/pybricks/support/issues/1095#issuecomment-1783899701) for?

The picture is of a logic analyzer attached to the Bluetooth pins on the MCU using really tiny clips that can clip on the the SMD leads. [link](https://www.pomonaelectronics.com/products/test-clips/micro-smd-grabber-test-clips-holding-rods-and-kit)

> More generally: what did you use to bootstrap pybricks on the board?
> Just UART-based debugging?

yup

**Reply by massimiliano-mantione** [reply]
> > More generally: what did you use to bootstrap pybricks on the board?
> > Just UART-based debugging?
> 
> yup

Which is amazing!

Apart from reverse-engineering the pybricks source code, is there some info on how to produce a firmware that can be loaded by the technic hub "bluetooth firmware flasher"?
Like, memory layout, entry point address...

And, for UART-based debugging, the best (or only?) way is to create an adapter cable and attach one of the ports to a serial adapter connected to a PC, right?

Again, apart from the source code, is there some recap of the technic hub peripherals and at which CPU GPIO pins they are attached?

**Reply by dlech** [reply]
No, there isn't really anywhere else to find this info other than the source code. Most of the info you are asking about is either in the linker script (`lib/pbio/platform/<name>/platform.ld`) or in the `lib/pbio/platform/<name>/platform.c` file.

And yeah, you will have to disable one of the I/O ports to be able to use the UART there and make your own cable.
